### PR TITLE
react: poc of async room initialisation

### PR DESCRIPTION
**Pull request by AndyTWF**
### Context

* Provide name of JIRA ticket that this PR is related to (if applicable).
    * e.g. [CHA-101] - this will automatically link the PR to the JIRA ticket.
* Provide names of any related DRs (if applicable).
* Provide links to any related PRs

### Description

* Describe the changes introduced in this PR
* Provide relevant context or considerations for reviewers.
    * Such as why the changes are necessary,
      or how they affect the system.

### Checklist

* [ ] QA'd by the author.
* [ ] Unit tests created (if applicable).
* [ ] Integration tests created (if applicable).
* [ ] Follow coding style guidelines found [here](https://github.com/ably/engineering/tree/main/best-practices).
* [ ] TypeDoc updated (if applicable).
* [ ] (Optional) Update documentation for new features.
* [ ] Browser tests created (if applicable).
* [ ] In repo demo app updated (if applicable).

### Testing Instructions (Optional)

* Explain how to test the changes in this PR.
* Provide specific steps or commands to execute.

[CHA-101]: https://ably.atlassian.net/browse/CHA-101?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/core/rooms.ts [hunk 1]
```diff
@@ -67,30 +75,45 @@ export class DefaultRooms implements Rooms {
...
+        this._logger.debug('Rooms.get(); returning new room', { roomId, opNumber: this.opNumber++ });
+        return entry.promise;
+      }
 
-    return room;
+      this._logger.debug('Rooms.get(); waiting for releasing to finish', { roomId: this.opNumber++ });
+      await releasing.promise;
+      this._logger.debug('Rooms.get(); releasing finished', { roomId: this.opNumber++ });
+      return this.get(roomId, options);
+    }, 1);
```
Comment (sacOO7) [inline src/core/rooms.ts:116]:
```suggestion
    }, RoomsOperationPrecendence.Get);
```

File: src/core/rooms.ts [hunk 2]
```diff
@@ -104,49 +127,64 @@ export class DefaultRooms implements Rooms {
...
+        this._logger.debug('Rooms.release(); room released', { roomId, opNumber: this.opNumber++ });
+        this._releasing.delete(roomId);
+      });
+      this._logger.debug('Rooms.release(); setting releasing', { roomId, opNumber: this.opNumber++ });
+      this._releasing.set(roomId, { count, promise: releasingPromise });
 
-    return releasedPromise;
+      // At this point, we might come back after the release has finished, so see if we're still releasing
+      return releasingPromise;
+    }, 2);
```
Comment (sacOO7) [inline src/core/rooms.ts:188]:
```suggestion
    }, RoomsOperationPrecendence.Release);
```

File: src/core/rooms.ts [hunk 3]
```diff
@@ -44,15 +46,40 @@ export interface Rooms {
...
  * Manages the chat rooms.
  */
 export class DefaultRooms implements Rooms {
   private readonly _realtime: Ably.Realtime;
   private readonly _chatApi: ChatApi;
   private readonly _clientOptions: NormalizedClientOptions;
-  private readonly _rooms: Map<string, DefaultRoom> = new Map<string, DefaultRoom>();
-  private readonly _releasing = new Map<string, { count: number; promise: Promise<void> }>();
+  private readonly _rooms: Map<string, RoomMapEntry> = new Map<string, RoomMapEntry>();
+  private readonly _releasing = new Map<string, { promise: Promise<void> }>();
```
Comment (sacOO7) [inline src/core/rooms.ts:82]: Can we add type alias for `Promise<void>`

**Comment by AndyTWF** [comment]
Superseded by https://github.com/ably/ably-chat-js/pull/387

**Pull request closed without merging**